Sources/Prelude/Refinements.swift

```diff
@@ -0,0 +1,136 @@
+    ///
+    /// - Parameter value: the proposed underlying value
+    /// - Throws: a `RefinementError` describing any failure
+    public init(_ value: A) throws {
+        guard R.isValid(value) else {
+            throw RefinementError("the value “\(value)” doesn’t satisfy the refinement “\(R.self)”")
+        }
+        self.value = value
+    }
+}
+
+extension Refined: Equatable where A: Equatable { }
+
+extension Refined: Hashable where A: Hashable { }
+
+public extension Refinement {
+    /// convenient way initialize a value of type `Refined<RefinedType, Self>` given a proposed value. This is equivalent to calling `try? Refined<RefinedType, Self>.init(value)`, but much easier to type
+    ///
+    /// - Parameter value: the proposed value
+    /// - Returns: a value of type `Refined<RefinedType, Self>`, or nil
+    static func of(_ value: RefinedType) -> Refined<RefinedType, Self>? {
+        return try? .init(value)
+    }
+}
+
+// MARK: - Both
+
+/// a refinement that can be used to combine any two other refinements (`L` and `R`), such that they both must pass
+public enum Both<L: Refinement, R: Refinement>: Refinement where L.RefinedType == R.RefinedType {
+    public typealias RefinedType = L.RefinedType
+    public static func isValid(_ value: RefinedType) -> Bool {
+        return L.isValid(value) && R.isValid(value)
+    }
+}
+
+/// given a value that is refined by two rules (`L` and `R`), produce a value that is refined only by the rule on the left (`L`). This function will always succeed (the underlying value has already passed both rules)
+///
+/// - Parameter refined: a value that is refined by two rules
+/// - Returns: the same value, refined only by the rule on the left
+public func left<A, L, R>(_ refined: Refined<A, Both<L, R>>) -> Refined<A, L> {
+    guard let result = L.of(refined.value) else {
+        preconditionFailure("when calling `left(_:)`: the value “\(refined.value)” didn’t satisfy the refinement “\(L.self)”. This shouldn’t happen. Was the value mutated after refinements had already been checked?")
+    }
+    return result
+}
+
+/// given a value that is refined by two rules (`L` and `R`), produce a value that is refined only by the rule on the right (`R`). This function will always succeed (the underlying value has already passed both rules)
+///
+/// - Parameter refined: a value that is refined by two rules
+/// - Returns: the same value, refined only by the rule on the right
+public func right<A, L, R>(_ refined: Refined<A, Both<L, R>>) -> Refined<A, R> {
+    guard let result = R.of(refined.value) else {
+        preconditionFailure("when calling `right(_:)`: the value “\(refined.value)” didn’t satisfy the refinement “\(R.self)”. This shouldn’t happen. Was the value mutated after refinements had already been checked?")
+    }
+    return result
+}
```

Comment:
having left and right be free functions feels a bit weird here to me. Maybe it's possible to add these to `Both` and `OneOf`, or to `Refined` when the Refinement is Both or OneOf?

Reply:
Yup, agreed. I’ll take a shot at making these methods… somewhere. fwiw, the clash of `R` type variables, which you mention below, was one reason I arrived at paralysis about where these should live. 👍

Reply:
These don’t work super well on `Refined` due to type inference issues (you have to specify the parameters a lot). I also think that having them on `Both` or `OneOf` is not ideal, because they’re not really factory functions of those types (for example, the `left` that takes a `Both` is actually trying to un-`Both` the value).

What I have locally right now is a caseless enum namespace called `Refine`, where these live, similar to what Promise does (I think it’s called `Promises`?).

That works! But it introduces another very-similarly-named type for kinda no reason (it’d give us `Refined`, `Refinement`, and `Refine`), and I’m not sure if that is more or less confusing than just having free functions.

Anyway, I wanted to re-propose that we leave these as free functions. There is a small precedent of there being free functions in `Prelude` already.

Reply:
That works for me! Thanks for looking into it, even if it proved impractical